packages/jest-jasmine2/src/tree_processor.js

```diff
@@ -43,40 +43,42 @@ export default function treeProcessor(options: Options) {
     return parentEnabled || runnableIds.indexOf(node.id) !== -1;
   }
 
-  return queueRunnerFactory({
-    onException: error => tree.onException(error),
-    queueableFns: wrapChildren(tree, isEnabled(tree, false)),
-    userContext: tree.sharedUserContext(),
-  });
-
-  function executeNode(node, parentEnabled) {
+  function getNodeHandler(node: TreeNode, parentEnabled: boolean) {
     const enabled = isEnabled(node, parentEnabled);
-    if (!node.children) {
-      return {
-        fn(done) {
-          node.execute(done, enabled);
-        },
-      };
-    }
-    return {
-      async fn(done) {
-        nodeStart(node);
-        await queueRunnerFactory({
-          onException: error => node.onException(error),
-          queueableFns: wrapChildren(node, enabled),
-          userContext: node.sharedUserContext(),
-        });
-        nodeComplete(node);
-        done();
-      },
+    return node.children
+      ? getNodeWithChildrenHandler(node, enabled)
+      : getNodeWithoutChildrenHandler(node, enabled);
+  }
+
+  function getNodeWithoutChildrenHandler(node: TreeNode, enabled: boolean) {
+    return function fn(done: (error?: any) => void = () => {}) {
+      node.execute(done, enabled);
+    };
+  }
+
+  function getNodeWithChildrenHandler(node: TreeNode, enabled: boolean) {
+    return async function fn(done: (error?: any) => void = () => {}) {
+      nodeStart(node);
+      await queueRunnerFactory({
+        onException: error => node.onException(error),
+        queueableFns: wrapChildren(node, enabled),
```

Comment:
After the refactor it's being called lazily in the `treeHandler` for nodes with children.